pkg/controller/direct/sql/sqlinstance_controller.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	api "google.golang.org/api/sqladmin/v1beta4"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/sql/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/k8s"
+	"github.com/googleapis/gax-go/v2"
+)
+
+const ctrlName = "sqlinstance-controller"
+
+func init() {
+	registry.RegisterModel(krm.SQLInstanceGVK, newSQLInstanceModel)
+}
+
+func newSQLInstanceModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &sqlInstanceModel{config: config}, nil
+}
+
+type sqlInstanceModel struct {
+	config *config.ControllerConfig
+}
+
+var _ directbase.Model = &sqlInstanceModel{}
+
+type sqlInstanceAdapter struct {
+	projectID  string
+	resourceID string
+
+	desired *krm.SQLInstance
+	actual  *api.DatabaseInstance
+
+	sqlOperationsClient *api.OperationsService
+	sqlInstancesClient  *api.InstancesService
+	sqlUsersClient      *api.UsersService
+}
+
+var _ directbase.Adapter = &sqlInstanceAdapter{}
+
+func (m *sqlInstanceModel) AdapterForObject(ctx context.Context, kube client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.SQLInstance{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	resourceID, err := refs.GetResourceID(u)
+	if err != nil {
+		return nil, err
+	}
+	obj.Spec.ResourceID = &resourceID
+
+	projectID, ok := u.GetAnnotations()[k8s.ProjectIDAnnotation]
+	if !ok {
+		projectID = u.GetNamespace()
```

Comment:
It looks like the `project-id` annotation is already getting applied to the resource, so we don't need to check the namespace also. I think this is due to the container annotation webhook. https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/24977c80ddda809ebccfc6fd32b9c835e61c6453/pkg/webhook/register.go#L139

Reply:
Ah great! Will it stay true after it is switched to the direct controller? Looks like this relies on the servicemapping.

Reply:
Discussed this offline, it will require some design to figure out best way to resolve project ids for direct controller resources. You offered to put together a doc w/ some background info, and I will fix / improve this in a follow-on PR.